openedx/core/lib/block_structure/transformers.py

```diff
@@ -111,39 +95,63 @@ def is_collected_outdated(cls, block_structure):
 
         return bool(outdated_transformers)
 
-    def _transform_optimized(self, transformers, block_structure):
+    def transform(self, block_structure):
+        """
+        The given block structure is transformed by each transformer in the
+        collection, in the order that the transformers were added.
+
+        All transformers that define transform_block_filter will be combined
+        into a single course tree traversal for faster performance.
         """
-        Transforms the given block_structure using the transform_block_filter
-        method from the given transformers.
+        combined_filters, full_transformers = self._combine_filters(block_structure)
+
+        # traverse once using all combinable filters
+        if combined_filters is not None:
+            block_structure.filter_topological_traversal(combined_filters)
+
+        # then transform the remainder
+        for transformer in full_transformers:
+            transformer.transform(self.usage_info, block_structure)
+
+        # Prune the block structure to remove any unreachable blocks.
+        block_structure._prune_unreachable()  # pylint: disable=protected-access
+
+    def _combine_filters(self, block_structure):
         """
-        if not transformers:
-            return
-
-        removal_filters = []
-        for removal in [t.transform_block_filter(self.usage_info, block_structure) for t in transformers]:
-            if isinstance(removal, list):
-                removal_filters.extend(removal)
-            else:
-                removal_filters.append(removal)
-
-        combined_filters = reduce(
-            lambda t1_filter, t2_filter: self._filter_chain(t1_filter, t2_filter),
-            removal_filters,
-            block_structure.create_universal_filter()
+        Attempts to combine transform_block_filters for all transformers.
+
+        Returns a tuple of the combined filter, and a list of transformers that
+        do not support this opimization. If no transformers support filtering,
+        combined_filter will be None.
+        """
+        no_filter_transformers = []
+        combined_filter, num_filters = block_structure.create_universal_filter(), 0
+
+        for transformer in self._transformers:
+            try:
+                combined_filter = self.filter_chain(
+                    combined_filter,
+                    transformer.transform_block_filter(self.usage_info, block_structure)
+                )
+                num_filters += 1
+            except NotImplementedError:
```

Comment:
This is the key to what I was thinking - instead of using an explicit subclass, or `_is_block_filter_implemented`, we try to get the filter from every transformer and handle the case where that method is not defined.